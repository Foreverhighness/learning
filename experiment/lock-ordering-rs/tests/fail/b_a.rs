use lock_ordering_rs::example::*;
use lock_ordering_rs::{OrderedMutex, new_lock_context};

#[expect(unused, reason = "testing")]
fn fail(a: OrderedMutex<A, ()>, b: OrderedMutex<D, ()>) {
    let mut ctx = new_lock_context();
    let (b, mut ctx) = b.lock(&mut ctx).unwrap();
    let (a, mut ctx) = a.lock(&mut ctx).unwrap();
}

fn main() {}
