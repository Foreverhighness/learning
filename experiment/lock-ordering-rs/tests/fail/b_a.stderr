error[E0277]: the trait bound `lock_ordering_rs::example::A: LockAfter<lock_ordering_rs::example::D>` is not satisfied
 --> tests/fail/b_a.rs:8:31
  |
8 |     let (a, mut ctx) = a.lock(&mut ctx).unwrap();
  |                          ---- ^^^^^^^^ the trait `LockAfter<lock_ordering_rs::example::D>` is not implemented for `lock_ordering_rs::example::A`
  |                          |
  |                          required by a bound introduced by this call
  |
  = help: the trait `LockAfter<lock_ordering_rs::example::D>` is not implemented for `lock_ordering_rs::example::A`
          but trait `LockAfter<Unlocked>` is implemented for it
  = help: for that trait implementation, expected `Unlocked`, found `lock_ordering_rs::example::D`
  = note: required for `lock_ordering_rs::example::D` to implement `LockBefore<lock_ordering_rs::example::A>`
note: required by a bound in `OrderedMutex::<Z, T>::lock`
 --> src/lib.rs
  |
  |     pub fn lock<A>(&self, _: &mut LockContext<A>) -> LockResult<Z, MutexGuard<T>>
  |            ---- required by a bound in this associated function
  |     where
  |         A: LockBefore<Z>,
  |            ^^^^^^^^^^^^^ required by this bound in `OrderedMutex::<Z, T>::lock`
